File: TOI neiroseti (python)/01/01.py
from decimal import *
getcontext().prec = 80 # чем больше знаков после запятой, тем длиннее строку можно закодировать

class Sector:
    char: str
    left: Decimal
    right: Decimal

def SetSectors(str, chars):
    # частота появления каждого уникального символа в строке
    freq = [[0 for k in range(2)] for j in range(len(chars))]
    i = 0
    for char in chars:
       freq[i][0] = char
       freq[i][1] = str.count(char) / len(str)
       i += 1
    # распределение границ секторов символов
    sector = [Sector() for i in range (len(freq))]
    left = Decimal(0)
    for i in range (0, len(freq)):
        sector[i].left = left
        sector[i].right = left + Decimal(freq[i][1])
        sector[i].char = freq[i][0]
        left = Decimal(sector[i].right)
    return sector

def ArithmeticEncode(str, sector):
    left = 0
    right = 1
    for i in range (len(str)):
        charSector = [x for x in sector if x.char == str[i]]
        newLeft = left + (right-left) * charSector[0].left
        newRight = left + (right-left) * charSector[0].right
        left = newLeft
        right = newRight
    return (left + right) / 2

def ArithmeticDecode(sector, chars, code, strLength):
    str = ""
    for i in range (strLength):
        for j in range (len(chars)):
            if (code >= sector[j].left and code < sector[j].right):
                str += sector[j].char
                code = (code - sector[j].left) / (sector[j].right - sector[j].left)
                break
    return str

print("Введите любую строку: ")
str = input()
chars = set(str) # сет каждый раз меняет порядок символов
sector = SetSectors(str, chars)
code = ArithmeticEncode(str, sector)
print("Закодированная строка:", code)
decodedStr = ArithmeticDecode(sector, chars, code, len(str))
print("Декодированная строка:", decodedStr)
if (str != decodedStr): print("Необходимо увеличить число знаков после запятой. Текущее число:", getcontext().prec)

File: TOI neiroseti (python)/01/01.cs
using System;
using System.Collections.Generic;
using System.Linq;
// (1) найдём частоту встречаемости каждого символа:
double[] FindFrequency(string str, char[] symbols)
{
    var frequency = new List<double>(); // список частот встречаемости
    foreach (char c in symbols)
    {
        var match = str.Where(x => x == c); // находим одинаковые символы
        frequency.Add((double)match.Count() / str.Length); // считаем частоту встречаемости каждого символа
    }
    return frequency.ToArray(); // возвращаем список в виде массива
}
// определим отрезки для каждого символа (мы уже имеем символы и их частоту):
Section[] DefineSection(char[] symbols, double[] frequency)
{
    Section[] section = new Section[symbols.Length]; // отрезок
    double l = 0; // левая граница отрезка
    for (int i = 0; i < symbols.Length; i++)
    {
        section[i].Left = l; // берём левую границу отрезка символа (сначала - от 0 до 1)
        section[i].Right = l + frequency[i]; // и правую, которая есть левая граница + частота символа
        section[i].Symbol = symbols[i]; // сохраняем сам символ
        l = section[i].Right; // переходим на следующий отрезок, что правее
    }
    return section; // сохраняем отрезок: символ, левая и правая границы
}
// АРИФМЕТИЧЕСКОЕ КОДИРОВАНИЕ:
// получаем кодируемое сообщение, его символы и их частоту
double ArithmeticEncoding(char[] symbols, double[] frequency, string str)
{
    // определяем отрезки:
    var section = DefineSection(symbols, frequency);
    // берём полный отрезок (от 0 до 1)
    double left = 0, right = 1;
    // проходимся по всей длине кодируемого сообщения:
    for (int i = 0; i < str.Length; i++)
    {
        int symb = Array.IndexOf(symbols, str[i]); // каждый символ сообщения
                                                   // делим отрезок:
        double newRight = left + (right - left) * section[symb].Right;
        double newLeft = left + (right - left) * section[symb].Left;
        // определяем следующий отрезок (который будем делить в след. раз)
        left = newLeft;
        right = newRight;
    }
    return (left + right) / 2; // возвращает длину отрезка
}
// ДЕКОДИРОВАНИЕ:
// получаем закодированное сообщение, длину отрезка, символы и их частоты
string ArithmeticDecoding(char[] symbols, double[] frequency, double code,
int n)
{ // определяем отрезки:
    var section = DefineSection(symbols, frequency);
    string str = ""; // берём строку, сначала пустую, далее она станет нашим ответом
    for (int i = 0; i < n; i++) // проходимся по всей длине отрезка
    {
        for (int j = 0; j < symbols.Length; j++) // проходимся по длине символа
        {
            // если закодированное сообщение находится в отрезке символа, то...
            if (code >= section[j].Left && code < section[j].Right)
            {
                str += section[j].Symbol; // прибовляем к строке символ
                                          //вычитаем из закодированного сообщения проверенные отрезки
                code = (code - section[j].Left) / (section[j].Right -
                section[j].Left);
                break;
            }
        }
    }
    return str; //выводим декодированное сообщение
}
Console.WriteLine("Write a string to encode: "); //запросим сообщение для кодирования:
string str = Console.ReadLine();
var symbols = str.Distinct().ToArray();
var frequency = FindFrequency(str, symbols);
var code = ArithmeticEncoding(symbols, frequency, str);
var decode = ArithmeticDecoding(symbols, frequency, code, str.Length);
//выводим результата и проверку:
Console.WriteLine("Encoded string: " + code);
Console.WriteLine("Decoded string: " + decode);
if (decode == str)
{
    Console.WriteLine("Success! Congratulations!");
}
else
{
    Console.WriteLine("Something went wrong");
}
//запись структуры (отрезков)
record struct Section(char Symbol, double Left, double Right);